url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/MinimizationOperations.java

```diff
@@ -0,0 +1,420 @@
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package net.pocorall.automaton;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Set;
+
+import static net.pocorall.automaton.BasicOperations.*;
+import static net.pocorall.automaton.SpecialOperations.*;
+
+/**
+ * Operations for minimizing automata.
+ */
+final public class MinimizationOperations {
+
+	private MinimizationOperations() {
+	}
+
+	private static boolean statesAgree(Transition[][] transitions, boolean[][] mark, int n1, int n2) {
+		Transition[] t1 = transitions[n1];
+		Transition[] t2 = transitions[n2];
+		for (int k1 = 0, k2 = 0; k1 < t1.length && k2 < t2.length; ) {
+			if (t1[k1].max < t2[k2].min)
+				k1++;
+			else if (t2[k2].max < t1[k1].min)
+				k2++;
+			else {
+				int m1 = t1[k1].to.number;
+				int m2 = t2[k2].to.number;
+				if (m1 > m2) {
+					int t = m1;
+					m1 = m2;
+					m2 = t;
+				}
+				if (mark[m1][m2])
+					return false;
+				if (t1[k1].max < t2[k2].max)
+					k1++;
+				else
+					k2++;
+			}
+		}
+		return true;
+	}
+
+	private static void addTriggers(Transition[][] transitions, ArrayList<ArrayList<HashSet<IntPair>>> triggers, int n1, int n2) {
+		Transition[] t1 = transitions[n1];
+		Transition[] t2 = transitions[n2];
+		for (int k1 = 0, k2 = 0; k1 < t1.length && k2 < t2.length; ) {
+			if (t1[k1].max < t2[k2].min)
+				k1++;
+			else if (t2[k2].max < t1[k1].min)
+				k2++;
+			else {
+				if (t1[k1].to != t2[k2].to) {
+					int m1 = t1[k1].to.number;
+					int m2 = t2[k2].to.number;
+					if (m1 > m2) {
+						int t = m1;
+						m1 = m2;
+						m2 = t;
+					}
+					if (triggers.get(m1).get(m2) == null)
+						triggers.get(m1).set(m2, new HashSet<IntPair>());
+					triggers.get(m1).get(m2).add(new IntPair(n1, n2));
+				}
+				if (t1[k1].max < t2[k2].max)
+					k1++;
+				else
+					k2++;
+			}
+		}
+	}
+
+	private static void markPair(boolean[][] mark, ArrayList<ArrayList<HashSet<IntPair>>> triggers, int n1, int n2) {
+		mark[n1][n2] = true;
+		if (triggers.get(n1).get(n2) != null) {
+			for (IntPair p : triggers.get(n1).get(n2)) {
+				int m1 = p.n1;
+				int m2 = p.n2;
+				if (m1 > m2) {
+					int t = m1;
+					m1 = m2;
+					m2 = t;
+				}
+				if (!mark[m1][m2])
+					markPair(mark, triggers, m1, m2);
+			}
+		}
+	}
+
+	private static <T> void initialize(ArrayList<T> list, int size) {
+		for (int i = 0; i < size; i++)
+			list.add(null);
+	}
+
+	/**
+	 * Minimizes the given automaton using Huffman's algorithm.
+	 */
+	public static void minimizeHuffman(LinkedAutomaton a) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 136-158 into a separate method.